libmamba/src/core/url.cpp

```diff
@@ -15,6 +17,152 @@
+
+        private:
+
+            pointer m_handle = nullptr;
+        };
+
+        /**
+         * A RAII wrapper for string mananged by Curl.
+         *
+         * String can possibly be null, or zero-lenght, depending on the data returned by Curl.
+         */
+        class CurlStr
+        {
+            using value_type = char;
+            using pointer = value_type*;
+            using const_pointer = const value_type*;
+            using input_pointer = value_type**;
+            using size_type = int;
+
+        public:
+
+            explicit CurlStr() = default;
+            explicit CurlStr(pointer data);
+            explicit CurlStr(pointer data, size_type len);
+            ~CurlStr();
+
+            [[nodiscard]] auto raw_input() -> input_pointer;
+            [[nodiscard]] auto raw() const -> const_pointer;
+            [[nodiscard]] auto raw() -> pointer;
+
+            [[nodiscard]] auto str() const -> std::optional<std::string_view>;
+
+        private:
+
+            pointer m_data = nullptr;
+            // Only meaningful when > 0, otherwise, assume null terminated string
+            size_type m_len = -1;
+
+            void free_data_if_present();
+        };
+
+        CurlEasyHandle::CurlEasyHandle()
+        {
+            m_handle = ::curl_easy_init();
+            if (m_handle == nullptr)
+            {
+                throw std::runtime_error("Could not create CURL handle");
+            }
+        }
+
+        CurlEasyHandle::~CurlEasyHandle()
+        {
+            ::curl_easy_cleanup(m_handle);
+        }
+
+        auto CurlEasyHandle::raw() const -> const_pointer
+        {
+            return m_handle;
+        }
+
+        auto CurlEasyHandle::raw() -> pointer
+        {
+            return m_handle;
+        }
+
+        CurlStr::CurlStr(pointer data)
+            : m_data(data)
+        {
+        }
+
+        CurlStr::CurlStr(pointer data, size_type len)
+            : m_data(data)
+            , m_len(len)
+        {
+        }
+
+        CurlStr::~CurlStr()
+        {
+            free_data_if_present();
+        }
+
+        auto CurlStr::raw_input() -> input_pointer
+        {
+            free_data_if_present();
+            return &m_data;
+        }
+
+        auto CurlStr::raw() const -> const_pointer
+        {
+            return m_data;
+        }
+
+        auto CurlStr::raw() -> pointer
+        {
+            return m_data;
+        }
+
+        auto CurlStr::str() const -> std::optional<std::string_view>
+        {
+            if (m_data)
+            {
+                if (m_len > 0)
+                {
+                    return { { m_data, static_cast<std::size_t>(m_len) } };
+                }
+                else
+                {
+                    return { { m_data } };
+                }
+            }
+            return std::nullopt;
+        }
+
+        void CurlStr::free_data_if_present()
+        {
+            if (m_data)
```

Comment:
This is not required, calling free (or ant curl-equivalent) on a null pointer is ok (it's a noop)